Suggested improvements

mmrwoods opened the pull request:
* Remove --column from ag and rg opts, unnecessary
* Rename ag.vim autoload script to grep.vim
* Use smart case or ignore case with grep commands
* Increase ASYNC_STEP to 10000, better performance
* Add syntax highlighting for filetypes known to vim
* Add support for fdfind (fd binary on Debian/Ubuntu)
* Make FuzzyFiles respect .gitignore by default
* Rename g:files_respect_gitignore to g:fuzzyy_files_respect_gitignore
* Use git-grep as fallback for FuzzyGrep when no ag or rg found
* Ignore binaries when using grep, git-grep, findstr
* Rename fuzzyy_files_ignore_{file,dir} to fuzzyy_files_exclude_{file,dir}
* Change default match hl from CurSearch to Special
* Fix cursor restore in MacVim (don't change &guicursor, change hi-Cursor instead)
* Fix FuzzyCommands for commands with arguments
* Support rg for finding files when no fd installed
* Rename enable_fuzzyy_keymaps to fuzzyy_enable_mapping
* Remove default value for g:fuzzyy_buffers_exclude, unnecessary
* Add custom fuzzyy highlight groups with defaults
* Fudge smart-case for grep, git-grep, findstr
* Fix fallback to git-grep, git-ls-files after chdir
* Add g:fuzzyy_grep_exclude_{file,dir}, similar to g:fuzzyy_files_exclude_{file,dir}
* Prefer rg to ag and fd for FuzzyGrep and FuzzyFiles (much faster than ag, almost as fast as fd, one less dependency)
* Include hidden files by default, with option to disable
* Add option to follow symlinks, disabled by default
* Allow customisation of files and grep commands (ripgrep only)
* Rename `:FuzzyyMRUFiles` to `:FuzzyyMru` (shorter, easier to read)
* Replace `g:fuzzyy_mru_project_only` with new `:FuzzyyMruCwd` command
* Add fewer leader mappings by default, and only those with obvious mnemonics
* Use preview window with :FuzzyHelps
* ~Rename `:FuzzyInBuffer` to `:FuzzySearch` (it's a fuzzy version of Vim's search command)~
* Use preview window with :FuzzyInBuffer

TODO: 
* support more custom highlighting options, results cursorline, individual popup borders etc.
* add ctrl+{s,t,v} key maps for FuzzyGrep, like FuzzyFiles etc.

mmrwoods commented:
Hi @Donaldttt,

First, thank you for Fuzzyy, it's a great plugin, I love the aim to have something that just works out of the box, that's exactly what I was looking for after using fzf.vim for years and leaderf for a while recently. 

I think it's time I asked for your feedback on some of these suggested changes before going much further. They are mostly either minor bug fixes or changes aiming to improve the out of the box experience so it "just works" as expected for new users (as expected being obviously highly subjective). They are driven by my personal use over a couple of weeks and my thoughts on how this should work as a plugin I would recommend to relative newcomers to Vim (I have recommended it btw!).

If you have a little time could you take a look at let me know what you think. I don't expect you to agree with every change, but I hope I have explained each one adequately and you can understand why I've made it, if not, please let me know and I'll clarify.

Thanks!

Donaldttt commented:
Hi, thank you for all the work you have done. I have been very busy lately. I will take a look of these change soon.

mmrwoods commented on autoload/fuzzy/ag.vim:
```diff
@@ -124,9 +124,14 @@ def AgJobStart(pattern: string)
 enddef
 
+def ErrCb(channel: channel, msg: string)
+    echoerr msg
```
This causes problems with `ag` and `--max-count` as it emits an error, fixed in later commit by adding `--silence` to `ag` options (doesn't silence critical errors like invalid options)

Donaldttt commented:
Thank you for the great work you have done! I check and I think most of the improvements are gread, but recently I don't have time to test and merge them. Would you like to be a collaborator for this project? 

For the changes:

Increase ASYNC_STEP to 10000, better performance
- this may depends on the spec of your machine. Maybe it would be better if it is an option.

Rename :FuzzyyMRUFiles to FuzzyyMru (shorter, easier to read)
(and other renamings)
- I agree, many naming can be improved. However, since they are breaking changes, maybe it would be better the old command can still be there for a while.

mmrwoods commented:
Hi @Donaldttt,

Thanks for offering to make me a collaborator, yes please, that would be great. I will not merge this PR as is, I still consider it a work in progress, but I'll wrap it up for final review soon.

As part of that wrapping up, I'll check all the breaking changes and see how best I can at least make sure users get notfiied with explanation telling users how to update their config/usage. I have already done this for many of the options, but not initially for the FuzzyMRU command as I thought it might be confusing to have both FuzzyMRU and FuzzyMru. Maybe it's worth the potential confusing in the short term.

RE the ASYNC_STEP, yep, you're right, it will depend on the machine, and probably should be configurable. I'm using an M1 Macbook air for most of my development, but I also test on a relatively puny Windows 10 AWS Workspace with 8G RAM. While `ASYNC_STEP = 10000` makes a huge difference to performance on my Macbook, it does actually make things perform a little worse on the Windows workspace. Setting to 5000 as a default works quite well on both, but you do still notice the performance degradation on the Macbook. As my Macbook is hardly top of the range these days, and this plugin is presumably targeted at developers who generally use pretty powerful machines, I think 10000 is probably still a reasonable default, but with the option to lower it if you're on an older machine. Leave that with me.

Thanks again

Mark

mmrwoods commented:
Just a thought... how about leaving the `vim9` as is, but creating a new `main` branch with these changes and then making that the default branch? This would be in addition to adding deprecation warnings etc., it would just allow people not ready to update their config to stick with the old code for a while.

mmrwoods commented:
@Donaldttt I've added an option to set the async step, made sure renamed commands are usable via their original name with a deprecation warning, and improved how option deprecations are handled.

Also renamed `FuzzyInBuffer` to `FuzzySearch`, because that was the best name I could come up with that was short, accurate (fuzzy version of Vim's search command) and had an obvious mnemonic to a default leader mapping that wouldn't conflict with other default leader mappings and would have the two keys to be typed close together (well, on a qwerty keyboard anyway).

And added some more default leader mappings, `<leader>fs` for `FuzzySearch` and `<leader>fr` for `FuzzyMruCwd` (think "fuzzy recent" or "find recent").

mmrwoods commented:
@Donaldttt `main` branch has been added and base/target of this PR changed to `main`. 

I don't plan to add any more commits to this PR, if you're happy with the latest changes we can merge to `main` and also change the default branch to `main`, leaving `vim9` for people who wish to remain pinned to the older code for a while.

Donaldttt commented:
These changes looked good to me!
